File: paleo_workbench/interchange/batch.py
"""Batch conversion service (I14): bounded, cancellable, failure-isolated.

This is *orchestration only* — every conversion is executed by the existing
adapter/export machinery, and results that belong in the catalog go through
it. There is deliberately no workflow DAG here (that is the Harness branch's
territory); this module is a stable service such a DAG could wrap.

Guarantees:
- bounded concurrency (default 2 workers, IO-bound background lane per
  ADR 0064)
- one failing item never aborts the batch (failure isolation) and never
  corrupts another item's output (atomic writes in the export path)
- cooperative cancellation at item boundaries and inside export checkpoints
- deterministic output naming (sorted stems, collision suffixes -2, -3, ...)
- pre-flight disk estimate before any work starts
"""

from __future__ import annotations

import time
from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass, field
from pathlib import Path
from typing import Callable

from paleo_workbench.interchange.contracts import (
    CancelToken,
    CancelledError,
    NULL_CANCEL,
    VerificationState,
)
from paleo_workbench.interchange.executor import ExportExecutor
from paleo_workbench.interchange.path_safety import sanitize_filename
from paleo_workbench.interchange.registry import InterchangeRegistry


@dataclass
class ConversionJob:
    source: Path
    target_format: str  # adapter format_id (e.g. "geojson", "csv", "raster", "flac3d_f3grid")
    target_name: str | None = None  # optional file name; default: <stem>.<ext>
    options: dict = field(default_factory=dict)


@dataclass
class BatchItemResult:
    source: str
    target: str = ""
    status: str = "pending"  # "converted" | "failed" | "skipped" | "cancelled"
    detail: str = ""
    duration_ms: int = 0
    verification_state: str = ""

    def to_dict(self) -> dict:
        return {
            "source": self.source,
            "target": self.target,
            "status": self.status,
            "detail": self.detail,
            "duration_ms": self.duration_ms,
            "verification_state": self.verification_state,
        }


@dataclass
class BatchResult:
    results: list[BatchItemResult] = field(default_factory=list)
    cancelled: bool = False
    total_duration_ms: int = 0
    estimated_disk_bytes: int = 0

    def summary(self) -> dict:
        counts: dict[str, int] = {}
        for item in self.results:
            counts[item.status] = counts.get(item.status, 0) + 1
        return {
            "total": len(self.results),
            **counts,
            "was_cancelled": self.cancelled,
            "total_duration_ms": self.total_duration_ms,
            "estimated_disk_bytes": self.estimated_disk_bytes,
        }

    def to_dict(self) -> dict:
        return {
            **self.summary(),
            "results": [r.to_dict() for r in self.results],
        }


ProgressFn = Callable[[int, int, str], None]


def _default_progress(done: int, total: int, current: str) -> None:
    return None


class BatchConversionService:
    def __init__(
        self,
        *,
        registry: InterchangeRegistry | None = None,
        max_workers: int = 2,
        catalog=None,
        project=None,
    ) -> None:
        self._registry = registry
        self._max_workers = max(1, min(int(max_workers), 4))
        self._catalog = catalog
        self._project = project

    def registry(self) -> InterchangeRegistry:
        if self._registry is None:
            from paleo_workbench.interchange.adapters import build_default_registry

            self._registry = build_default_registry()
        return self._registry

    # -- planning -----------------------------------------------------------
    def estimate(self, jobs: list[ConversionJob], *, output_dir: Path) -> tuple[int, list[str]]:
        """Return (estimated_disk_bytes, warnings) without converting."""
        total = 0
        warnings: list[str] = []
        for job in jobs:
            adapter = self._adapter_for(job.source, job.target_format)
            if adapter is None:
                warnings.append(f"无法识别: {job.source}")
                continue
            try:
                plan = adapter.plan_export(
                    job.source,
                    self._target_path(job, output_dir),
                    options=job.options,
                )
            except Exception as exc:
                warnings.append(f"{job.source}: {exc}")
                continue
            total += plan.estimated_bytes
        return total, warnings

    # -- execution ----------------------------------------------------------
    def convert(
        self,
        jobs: list[ConversionJob],
        *,
        output_dir: Path,
        cancel: CancelToken | None = None,
        progress: ProgressFn | None = None,
        verify: bool = True,
    ) -> BatchResult:
        cancel = cancel or NULL_CANCEL
        progress = progress or _default_progress
        progress_ok = True  # a failing callback must not lose the batch result

        def safe_progress(done: int, total: int, current: str) -> None:
            nonlocal progress_ok
            if not progress_ok:
                return
            try:
                progress(done, total, current)
            except Exception:
                progress_ok = False

        output_dir = Path(output_dir)
        output_dir.mkdir(parents=True, exist_ok=True)
        started = time.monotonic()
        result = BatchResult()
        result.estimated_disk_bytes, _ = self.estimate(jobs, output_dir=output_dir)

        # deterministic order: sort by (source path, target name)
        ordered = sorted(jobs, key=lambda j: (str(j.source), j.target_name or ""))
        ordered = self._dedupe_targets(ordered)

        cancel.checkpoint()
        total = len(ordered)
        if total == 0:
            return result
        done = 0

        def run_one(job: ConversionJob) -> BatchItemResult:
            item = BatchItemResult(source=str(job.source))
            item_start = time.monotonic()
            cancel.checkpoint()
            try:
                adapter = self._adapter_for(job.source, job.target_format)
                if adapter is None:
                    item.status = "skipped"
                    item.detail = "无适配器或能力不可用"
                    return item
                target = self._target_path(job, output_dir)
                plan = adapter.plan_export(job.source, target, options=job.options)
                item.target = str(target)
                executor = ExportExecutor(
                    registry=self.registry(),
                    work_dir=output_dir / ".work",
                    catalog=self._catalog,
                    project=self._project,
                )
                _, verification = executor.execute(plan, cancel=cancel, verify=verify)
                item.verification_state = verification.state.value
                # Only FAILED fails the item. With verify=False the output is
                # UNVERIFIED (recorded honestly), which the caller opted into.
                if verification.state is VerificationState.FAILED:
                    item.status = "failed"
                    item.detail = verification.detail or "输出未通过校验"
                else:
                    item.status = "converted"
                    if verification.state is VerificationState.UNVERIFIED and verify:
                        item.detail = verification.detail or "输出未验证（校验器异常）"
            except CancelledError:
                item.status = "cancelled"
                raise
            except Exception as exc:
                item.status = "failed"
                item.detail = str(exc)
            finally:
                item.duration_ms = int((time.monotonic() - item_start) * 1000)
            return item

        if self._max_workers == 1 or total == 1:
            for job in ordered:
                try:
                    result.results.append(run_one(job))
                except CancelledError:
                    result.results.append(BatchItemResult(
                        source=str(job.source), status="cancelled"))
                    if cancel.cancelled:
                        result.cancelled = True
                    # a job-internal CancelledError is failure isolation, not
                    # a batch stop — only a cancelled shared token stops us
                done += 1
                safe_progress(done, total, str(job.source))
                if result.cancelled:
                    for remaining in ordered[done:]:
                        result.results.append(BatchItemResult(
                            source=str(remaining.source), status="cancelled"))
                    break
        else:
            with ThreadPoolExecutor(max_workers=self._max_workers) as pool:
                futures = [pool.submit(run_one, job) for job in ordered]
                for job, future in zip(ordered, futures):
                    try:
                        result.results.append(future.result())
                    except CancelledError:
                        result.results.append(BatchItemResult(
                            source=str(job.source), status="cancelled"))
                        if cancel.cancelled:
                            result.cancelled = True
                    except Exception as exc:  # defensive: isolate everything
                        result.results.append(BatchItemResult(
                            source=str(job.source), status="failed", detail=str(exc)))
                    done += 1
                    safe_progress(done, total, str(job.source))
            if cancel.cancelled:
                result.cancelled = True
                for item in result.results:
                    if item.status == "pending":
                        item.status = "cancelled"
        result.total_duration_ms = int((time.monotonic() - started) * 1000)
        # order results deterministically regardless of completion order
        result.results.sort(key=lambda r: r.source)
        return result

    # -- helpers ------------------------------------------------------------
    def _dedupe_targets(self, ordered: list[ConversionJob]) -> list[ConversionJob]:
        """Deterministic collision-free names: <stem>-2.ext, -3.ext, ..."""
        from dataclasses import replace

        used: set[str] = set()
        result: list[ConversionJob] = []
        for job in ordered:
            base = self._target_path(job, output_dir=Path("."))
            candidate = base
            counter = 1
            while candidate.name.casefold() in used:
                counter += 1
                candidate = base.with_name(f"{base.stem}-{counter}{base.suffix}")
            used.add(candidate.name.casefold())
            result.append(replace(job, target_name=candidate.name))
        return result

    def _adapter_for(self, source: Path, target_format: str):
        adapter = self.registry().get(target_format)
        if adapter is not None and adapter.capability().export:
            return adapter
        return None

    def _target_path(self, job: ConversionJob, output_dir: Path) -> Path:
        adapter = self.registry().get(job.target_format)
        extension = adapter.extensions[0] if adapter and adapter.extensions else "bin"
        if job.target_name:
            name = sanitize_filename(job.target_name)
        else:
            name = f"{Path(job.source).stem}.{extension}"
        return output_dir / name

File: libs/interchange/src/batch.cpp
// Implementation of batch.hpp — branch-for-branch port of
// paleo_workbench/interchange/batch.py over the native model-adapter
// service.
#include <pwb/interchange/batch.hpp>

#include <algorithm>
#include <chrono>
#include <mutex>
#include <thread>

#include <pwb/interchange/path_safety.hpp>
#include <pwb/interchange/unicode.hpp>

namespace pwb::interchange {

namespace {

long long monotonic_ms_since(std::chrono::steady_clock::time_point start) {
    return std::chrono::duration_cast<std::chrono::milliseconds>(
               std::chrono::steady_clock::now() - start)
        .count();
}

}  // namespace

Json BatchItemResult::to_json() const {
    Json out = Json::object();
    out["source"] = source;
    out["target"] = target;
    out["status"] = status;
    out["detail"] = detail;
    out["duration_ms"] = duration_ms;
    out["verification_state"] = verification_state;
    return out;
}

Json BatchResult::summary() const {
    Json counts = Json::object();
    for (const auto& item : results) {
        counts[item.status] = counts.value(item.status, 0) + 1;
    }
    Json out = Json::object();
    out["total"] = static_cast<long long>(results.size());
    // Python {**counts} spread: the count keys land after "total".
    for (auto it = counts.begin(); it != counts.end(); ++it) {
        out[it.key()] = it.value();
    }
    out["was_cancelled"] = cancelled;
    out["total_duration_ms"] = total_duration_ms;
    out["estimated_disk_bytes"] = estimated_disk_bytes;
    return out;
}

Json BatchResult::to_json() const {
    Json out = summary();
    out["results"] = Json::array();
    for (const auto& item : results) {
        out["results"].push_back(item.to_json());
    }
    return out;
}

const ModelAdapter* BatchConversionService::adapter_for(
    const std::filesystem::path& source, const std::string& target_format) const {
    (void)source;
    const ModelAdapter* adapter = service_.adapter(target_format);
    if (adapter != nullptr && adapter->capability().can_export) {
        return adapter;
    }
    return nullptr;
}

std::filesystem::path BatchConversionService::target_path(
    const ConversionJob& job, const std::filesystem::path& output_dir) const {
    const ModelAdapter* adapter = service_.adapter(job.target_format);
    const std::string extension =
        (adapter != nullptr && !adapter->extensions.empty()) ? adapter->extensions.front()
                                                             : std::string("bin");
    std::string name;
    if (job.target_name.has_value()) {
        name = sanitize_filename(*job.target_name);
    } else {
        name = job.source.stem().string() + "." + extension;
    }
    return output_dir / name;
}

std::pair<long long, std::vector<std::string>> BatchConversionService::estimate(
    const std::vector<ConversionJob>& jobs,
    const std::filesystem::path& output_dir) const {
    long long total = 0;
    std::vector<std::string> warnings;
    for (const auto& job : jobs) {
        const ModelAdapter* adapter = adapter_for(job.source, job.target_format);
        if (adapter == nullptr) {
            warnings.push_back("无法识别: " + job.source.string());
            continue;
        }
        try {
            const ExportPlan plan =
                adapter->plan_export(job.source, target_path(job, output_dir), job.options);
            total += plan.estimated_bytes;
        } catch (const std::exception& exc) {
            warnings.push_back(job.source.string() + ": " + exc.what());
        }
    }
    return {total, warnings};
}

std::vector<ConversionJob> BatchConversionService::dedupe_targets(
    const std::vector<ConversionJob>& ordered,
    const std::filesystem::path& output_dir) const {
    // Deterministic collision-free names: <stem>-2.ext, -3.ext, ...
    std::vector<std::string> used;
    std::vector<ConversionJob> result;
    for (ConversionJob job : ordered) {
        std::string base = target_path(job, output_dir).filename().string();
        std::string candidate = base;
        int counter = 1;
        while (true) {
            const std::string folded = casefold_utf8(candidate);
            if (std::find(used.begin(), used.end(), folded) == used.end()) break;
            ++counter;
            const std::filesystem::path base_path(base);
            candidate = base_path.stem().string() + "-" + std::to_string(counter) +
                        base_path.extension().string();
        }
        used.push_back(casefold_utf8(candidate));
        job.target_name = candidate;
        result.push_back(std::move(job));
    }
    return result;
}

BatchResult BatchConversionService::convert(const std::vector<ConversionJob>& jobs,
                                            const std::filesystem::path& output_dir,
                                            const CancelToken& cancel,
                                            BatchProgressFn progress,
                                            bool verify) const {
    cancel.checkpoint();
    const auto started = std::chrono::steady_clock::now();
    BatchResult result;
    result.estimated_disk_bytes = estimate(jobs, output_dir).first;

    std::filesystem::create_directories(output_dir);

    // Deterministic order: sort by (source path, target name).
    std::vector<ConversionJob> ordered = jobs;
    std::sort(ordered.begin(), ordered.end(), [](const ConversionJob& a, const ConversionJob& b) {
        const std::string a_source = a.source.string();
        const std::string b_source = b.source.string();
        if (a_source != b_source) return a_source < b_source;
        return a.target_name.value_or("") < b.target_name.value_or("");
    });
    ordered = dedupe_targets(ordered, output_dir);

    // A failing callback must not lose the batch result.
    auto safe_progress = [&progress](int done, int total, const std::string& current) {
        if (!progress) return;
        try {
            progress(done, total, current);
        } catch (...) {
        }
    };

    auto run_one = [&](const ConversionJob& job) {
        BatchItemResult item;
        item.source = job.source.string();
        const auto item_start = std::chrono::steady_clock::now();
        try {
            cancel.checkpoint();
            const ModelAdapter* adapter = adapter_for(job.source, job.target_format);
            if (adapter == nullptr) {
                item.status = "skipped";
                item.detail = "无适配器或能力不可用";
                item.duration_ms = monotonic_ms_since(item_start);
                return item;
            }
            const std::filesystem::path target = target_path(job, output_dir);
            ExportPlan plan = adapter->plan_export(job.source, target, job.options);
            item.target = target.string();
            plan.target_path = target.string();
            const std::filesystem::path output = adapter->export_data(plan, cancel);
            ExportVerification verification =
                verify ? adapter->verify_output(output, plan)
                       : ExportVerification::unverified("verify=False（调用方选择跳过校验）");
            item.verification_state = std::string(to_string(verification.state));
            // Only FAILED fails the item. With verify=False the output is
            // UNVERIFIED (recorded honestly), which the caller opted into.
            if (verification.state == VerificationState::FAILED) {
                item.status = "failed";
                item.detail =
                    !verification.detail.empty() ? verification.detail : "输出未通过校验";
            } else {
                item.status = "converted";
                if (registration_ != nullptr && verification.ok()) {
                    try {
                        registration_->register_export(plan, output, verification);
                    } catch (const std::exception& exc) {
                        // 登记失败追加 warning 不改结果。
                        item.detail = std::string("输出登记失败: ") + exc.what();
                    }
                } else if (verification.state == VerificationState::UNVERIFIED && verify) {
                    item.detail = !verification.detail.empty()
                                      ? verification.detail
                                      : "输出未验证（校验器异常）";
                }
            }
        } catch (const CancelledError&) {
            item.status = "cancelled";
            item.duration_ms = monotonic_ms_since(item_start);
            throw;
        } catch (const std::exception& exc) {
            item.status = "failed";
            item.detail = exc.what();
        }
        item.duration_ms = monotonic_ms_since(item_start);
        return item;
    };

    const int total = static_cast<int>(ordered.size());
    int done = 0;
    if (max_workers_ == 1 || total <= 1) {
        for (const auto& job : ordered) {
            try {
                result.results.push_back(run_one(job));
            } catch (const CancelledError&) {
                BatchItemResult item;
                item.source = job.source.string();
                item.status = "cancelled";
                result.results.push_back(std::move(item));
                if (cancel.cancelled()) result.cancelled = true;
                // A job-internal CancelledError is failure isolation, not a
                // batch stop — only a cancelled shared token stops us.
            }
            ++done;
            safe_progress(done, total, job.source.string());
            if (result.cancelled) {
                // Everything after the shared-token cancellation is
                // cancelled without running.
                for (std::size_t i = done; i < ordered.size(); ++i) {
                    BatchItemResult item;
                    item.source = ordered[i].source.string();
                    item.status = "cancelled";
                    result.results.push_back(std::move(item));
                }
                break;
            }
        }
    } else {
        // Bounded worker pool; results collected in submission order.
        std::mutex result_mutex;
        std::vector<BatchItemResult> collected(ordered.size());
        std::vector<std::thread> workers;
        std::size_t next = 0;
        bool stop = false;
        for (int worker = 0; worker < max_workers_ && worker < total; ++worker) {
            workers.emplace_back([&] {
                for (;;) {
                    std::size_t index = 0;
                    {
                        std::lock_guard<std::mutex> guard(result_mutex);
                        if (stop || next >= ordered.size()) return;
                        index = next++;
                    }
                    BatchItemResult item;
                    try {
                        item = run_one(ordered[index]);
                    } catch (const CancelledError&) {
                        item.source = ordered[index].source.string();
                        item.status = "cancelled";
                        if (cancel.cancelled()) {
                            std::lock_guard<std::mutex> guard(result_mutex);
                            stop = true;
                        }
                    } catch (const std::exception& exc) {  // defensive: isolate everything
                        item.source = ordered[index].source.string();
                        item.status = "failed";
                        item.detail = exc.what();
                    }
                    {
                        std::lock_guard<std::mutex> guard(result_mutex);
                        collected[index] = std::move(item);
                    }
                }
            });
        }
        for (auto& worker : workers) worker.join();
        for (auto& item : collected) {
            result.results.push_back(std::move(item));
            ++done;
            // Progress after the fact keeps the ordering deterministic; the
            // current label mirrors the just-finished job.
            safe_progress(done, total, result.results.back().source);
        }
        if (cancel.cancelled()) {
            result.cancelled = true;
            for (auto& item : result.results) {
                if (item.status == "pending") item.status = "cancelled";
            }
        }
    }
    result.total_duration_ms = monotonic_ms_since(started);
    // Order results deterministically regardless of completion order.
    std::stable_sort(result.results.begin(), result.results.end(),
                     [](const BatchItemResult& a, const BatchItemResult& b) {
                         return a.source < b.source;
                     });
    return result;
}

}  // namespace pwb::interchange
